### ISSUE TITLE
WO-007: reporting and reward

### ISSUE THREAD

**Issue by Xocas12** [Xocas12/gosplan-env#12]
> **WO-007** - Phase 1 - Owner **MID-strong** - Difficulty **3/5** - Gate **G0**
> Card: [`workorders/WO-007.md`](https://github.com/Xocas12/gosplan-env/blob/main/workorders/WO-007.md)

## What this is

This card implements the incentive machinery: what happens when an enterprise files a report, what
happens when that report is audited, and what the enterprise is paid. Fourteen bodies across three
files - reporting and audit/penalty, the bonus schedule with its notch and the reward that is
exactly five terms and no more, the logged welfare and value metrics, and the cost-plus price fixed
point.

It exists because the reward is the entire experimental treatment. CONTRACT rule 4 fixes it term for
term: no shaping, no auxiliary reward, no running normalisation. If anything extra gets added here,
every downstream claim about emergence is worthless - the behaviour would have been paid for
directly.

## Depends on

- WO-003
- WO-004

The frozen suite this card must pass is WO-002's.

**Blocked by:** #8 (WO-003), #9 (WO-004)
This card may not be started until every issue above is closed (CONTRACT rule 12 / the loop in CONTRIBUTING.md).

## Read first (exhaustive whitelist)

- [`spec/spec.py`](https://github.com/Xocas12/gosplan-env/blob/main/spec/spec.py) - frozen signatures
  and docstrings of `process_reports`, `audit_and_penalise`, `bonus`, `reward_scale`,
  `enterprise_reward`, `val_measured`, `val_true`, `welfare_true`, `initial_prices`.
- [`gosplan/env/reporting.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/reporting.py) -
  **write target 1.** Contracts for `process_reports` and `audit_and_penalise` are already there; two
  bodies missing.
- [`gosplan/env/reward.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/reward.py) -
  **write target 2.** Contracts for `bonus`, `reward_scale`, `enterprise_reward`, `val_measured`,
  `val_true`, `welfare_true`, `padding_index`, `welfare_ratio`, `specification_gap`; nine bodies
  missing.
- [`gosplan/env/prices.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/prices.py) -
  **write target 3.** `LABOUR_COST`, `PRICE_PERTURBATION_SIGMA`, `N_PRICE_PERTURBATIONS`, and
  contracts for `initial_prices`, `recompute_prices`, `perturbed_price_vectors`; three bodies missing.
- [`gosplan/config.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/config.py) -
  `EnvConfig` and the INC / INFO / SUPPLY / TECH fields this card reads: `notch_height`,
  `notch_width`, `overfulfilment_slope`, `overfulfilment_cap`, `penalty_form`, `penalty_arg`,
  `penalty_scale`, `audit_noise`, `quality_measurability`, `holding_loss`, `inventory_cap_mult`,
  `report_max_ratio`, `request_max_multiple`, `price_markup`, `io_matrix`, `ces_alpha`, `ces_sigma`,
  `final_demand_share`, `sector_of`.
- [`gosplan/rng.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/rng.py) - `draw` and
  `PURPOSES`; purpose `auditnoise` is this card's.
- [`gosplan/env/state.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/state.py) -
  `State` and `EnterpriseAction` fields, and the inventory rules of PLAN section 2.11 that the record
  is the home of.
- [`gosplan/env/planner.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/planner.py) -
  `PlannerView` and `fulfilment_measure`, for the `rho` the reward keys on.
- [`CONTRACT.md`](https://github.com/Xocas12/gosplan-env/blob/main/CONTRACT.md) - rule 4 (REWARD
  TERMS) above all, plus rules 3, 6 (welfare blindness), 8 (bounds are results), 9, 12.
- `docs/ref_worked_example.md` - the hand-validated two-enterprise, two-sector trajectory (written by
  WO-002).
- [`pyproject.toml`](https://github.com/Xocas12/gosplan-env/blob/main/pyproject.toml) - ruff
  line-length 100, target py312.
- [`tests/unit/test_reporting.py`](https://github.com/Xocas12/gosplan-env/blob/main/tests/unit/test_reporting.py) -
  read-only (CONTRACT rule 2), from WO-002.
- [`tests/unit/test_reward.py`](https://github.com/Xocas12/gosplan-env/blob/main/tests/unit/test_reward.py) -
  read-only (CONTRACT rule 2), from WO-002.
- [`tests/unit/test_prices.py`](https://github.com/Xocas12/gosplan-env/blob/main/tests/unit/test_prices.py) -
  read-only (CONTRACT rule 2), from WO-002.

Read nothing else. CONTRACT rule 12.

## Write only

- `gosplan/env/reporting.py`
- `gosplan/env/reward.py`
- `gosplan/env/prices.py`

Creating, editing or deleting any other file is a CONTRACT rule 12 violation.

## Must pass

- [ ] `tests/unit/test_reporting.py` - **T-U8**: `positive_part` gives 0 for any under-report,
      `absolute` does not, `audited = False` gives 0. Plus: the audit compares the claim to stock;
      the holding loss is applied before `y` is added; the report is clipped to `rho_max`.
- [ ] `tests/unit/test_reward.py` - **T-U2**: `reward_scale(cfg) * B_cfg(1.1) == 1`; **T-U3**:
      `bonus` monotone in `rho`, discontinuous at 1 iff `w = 0`, continuous with continuous
      derivative iff `w > 0` and `rho_cap = inf`; **T-B6 components**: the reward equals the formula
      recomputed independently. Plus: `val`, `val_true`, and the CES welfare with the
      `sigma_c -> 1` limit equal to Cobb-Douglas.
- [ ] `tests/unit/test_prices.py` - the cost-plus fixed point converges and every price is positive.

```bash
uv run pytest tests/unit/test_reporting.py -q
uv run pytest tests/unit/test_reward.py -q
uv run pytest tests/unit/test_prices.py -q
uv run ruff check gosplan/env/reporting.py gosplan/env/reward.py gosplan/env/prices.py
uv run ruff format --check gosplan/env/reporting.py gosplan/env/reward.py gosplan/env/prices.py
```

## Implementation notes

Fourteen bodies across three files. Every formula is PLAN sections 2.8-2.10 verbatim.

### A. `gosplan/env/reporting.py`

**`process_reports(state, action, cfg) -> State`.** PLAN section 2.8:

```
S_i <- (1 - h) * S_i + y_i       # loss on the stock carried IN, THEN this period's y
R_i  = clip(rho_i_report, 0, rho_max) * T_i
```

**Order matters and is tested**: the loss applies to the carried stock, not to output just produced.
Stock above `S_max = cfg.tech.inventory_cap_mult * cap_i` is LOST, and the overflow is recorded so it
stays visible in the conservation identity T-U1 (PLAN section 2.11).

The function also stores `last_report_ratio`, `last_report` (the claim in units, kept because the
ratchet moves `T` later in the same period - PLAN section 2.5 runs REPORT at step 3 and TARGET at
step 6), and `request`, clipped to `r_max * need_ij` with `r_max = cfg.tech.request_max_multiple`. It
records whether the report sat at `rho_max`, which is what `BOUND_BINDING` (CONTRACT rule 8, test
T-B8) is computed from.

At this step the agent has already observed `S_i` and `y_i` exactly in Phase 1
(`self_obs_noise = 0`), so the report is a choice made under full knowledge of the truth.

**`audit_and_penalise(state, audited, cfg, t) -> Array`.** PLAN section 2.8:

```
S_hat_i   = S_i * exp(nu_i),  nu_i ~ N(0, sigma_aud**2)
            key = (seed_env, "auditnoise", t, i)
f_i       = max(0, R_i - S_hat_i) / T_i     if penalty_arg = positive_part      (Phase 1)
          = |R_i - S_hat_i| / T_i           if penalty_arg = absolute
Pen_i     = pen * f_i                       if penalty_form = proportional      (Phase 1)
          = pen * 1[f_i > 0]                if penalty_form = fixed
penalty_i = 1[audited_i] * Pen_i
```

**The audit compares the claim to stock on hand, never to the period's production.** That one choice
is what makes accumulated stock protect against audits, and it is why `penalty_arg` and `h` are the
mechanism parameters behind the held-out hidden-reserves phenomenon (PLAN section 4.2). The penalty
is in RATIO units - divided by `T_i` - per finding F9.

Draw the noise **unconditionally**, even at `sigma_aud = 0` where `exp(0) = 1` exactly, so the key
layout does not depend on the configuration; key-based draws are order-independent, so an unused key
costs nothing.

### B. `gosplan/env/reward.py`

**`bonus(rho, cfg) -> Array`.** PLAN section 2.8:

```
Lambda_w(x) = 1[x >= 0]                if w = 0        # strict >=, a true Heaviside
            = 1 / (1 + exp(-x / w))    if w > 0
B(rho)      = beta * Lambda_w(rho - 1) + s * clip(rho - 1, 0, rho_cap - 1)
```

`w = 0` is a genuine discontinuity and `B(1)` therefore **includes** `beta`. `rho_cap = inf` means no
cap at all and hence no kink: in that branch do not clip from above, i.e. the second term is
`s * max(rho - 1, 0)`.

Named configurations: notched `(w = 0, rho_cap = 1.2)`; smooth counterfactual `(w = 0.25,
rho_cap = inf)`; kink-only `(w = 0.25, rho_cap = 1.2)`. The smooth arm shares `beta` and `s` with the
notched arm - it is equal-parameter, not equal-expected-value.

**`reward_scale(cfg) -> float`.** `scale = 1 / B_cfg(rho_ref = 1.1)`, computed **analytically** from
the configuration and NEVER from running statistics (finding F9, CONTRACT rule 4). It makes the bonus
at 110 per cent fulfilment equal to 1 in every configuration, so sweeping `beta` changes the
economics - the notch relative to `pen` and `kappa` - and not the gradient magnitude.

**Card decision, because PLAN section 2.9.1 does not cover it:** when `beta = 0` and `s = 0` the
denominator is 0 and the scale is undefined; **raise `ValueError` naming both fields.** Do not return
`inf`, do not substitute `1.0`. If `tests/unit/test_reward.py` or `tests/unit/test_dp.py` expects
another behaviour, file an AMBIGUITY REPORT and stop.

**`enterprise_reward(state, cfg, phase, cost, penalty, trade_surplus) -> Array`.** PLAN section 2.9.1
and CONTRACT rule 4, exactly these terms and no others:

```
PRODUCE step k:   r_ik = - scale * c_ik
REPORT step:      r_i  =   scale * ( B(rho_i) - penalty_i + trade_surplus_i )
scale             = reward_scale(cfg)
```

`trade_surplus` is identically 0 in Phase 1; a `None` argument in the phase where a term does not
apply is treated as zeros, not as an error. `rho_i` is `fulfilment_measure(...) / T_i`.

**Metrics.** `val_measured(state, cfg) -> float` = `sum_i p_{s(i)} * R_i * q_hat_i`, with
`q_hat_i = 1 + mu * (qbar_i - 1)` (Phase 1: 1). `val_true(state, cfg) -> float` =
`sum_i p_{s(i)} * y_i * qbar_i`. `welfare_true(consumer, cfg) -> float`:

```
rho_ces   = (sigma_c - 1) / sigma_c
welfare_t = ( sum_j alpha_j * consumer_j**rho_ces )**(1 / rho_ces)
```

The `sigma_c -> 1` limit is Cobb-Douglas, `prod_j consumer_j**alpha_j`, and must be an **explicit
branch** (it is on the must-pass list). Consequence of the formula that must be produced **without a
warning or a NaN**: at `sigma_c < 1` (Phase 1 is 0.8, complements-leaning) `rho_ces` is negative, so
a single `consumer_j = 0` drives the index to 0 - a bundle missing a good is worth nothing. That is
the correct value, not an edge case to be regularised.

**`padding_index`, `welfare_ratio`, `specification_gap`** (PLAN section 2.9.4). Each ratio is formed
from **window means** - `mean_t val_measured_t / mean_t val_true_t`, not the mean of per-period ratios
- so that the three metrics are commensurable and match `W = mean_t welfare_t`. The measurement
window is PLAN section 4.4 (periods `t >= 2`) and **the caller supplies it**. `welfare_oracle <= 0`
or `val_oracle <= 0` is a solver or configuration failure and must **raise** rather than produce a
signed ratio. In Phase 1 the oracle denominators are the clearly labelled `W_truthful_max`
placeholder; this module does not compute them and does not silently substitute one.

### C. `gosplan/env/prices.py`

**`initial_prices(cfg) -> Array`.** PLAN section 2.10, the cost-plus fixed point

```
p_j = (1 + m) * (kappa_labour + sum_k a_jk * p_k)
```

with `m = cfg.supply.price_markup` and `kappa_labour = LABOUR_COST = 1.0`, a **normalisation rather
than a parameter**: `p` is homogeneous of degree 1 in it and every use of `p` in this design is a
ratio, so the scale cancels. Solve it directly - `(I - (1 + m) A) p = (1 + m) * kappa_labour * 1` -
or by iteration to convergence; either way return strictly positive prices.

Convergence needs `(1 + m) * sum_k a_jk < 1` for every row, which is why `EnvConfig.validate` rejects
`sum_k a_jk >= 1`; note in your report that the validation rule is the looser of the two (see the
WO-003 card) and **do not tighten it from here.**

**`recompute_prices(planner_io, cfg) -> Array`** is the same solver with `planner_io` in place of
`cfg.supply.io_matrix`, and it **must not read `cfg.supply.io_matrix`**: the whole point is that once
I-O drift moves the true coefficients while `State.planner_io` stays fixed, prices are computed from
a stale picture. Inert in Phase 1 (`price_lag = inf`). The recomputation SCHEDULE belongs to the step
machine (WO-009), not here.

**`perturbed_price_vectors(prices, seeds)` - OPEN ITEM, do not resolve it.** PLAN sections 2.9.4 and
7.5 require `p_j * exp(u_j)` with `u ~ N(0, PRICE_PERTURBATION_SIGMA**2)` at fixed seeds, but
CONTRACT rule 9 requires every draw under `gosplan/env/` to go through `gosplan.rng.draw`, and
`Purpose` has no value for a price perturbation. The two admissible resolutions - add a purpose at
the v1 freeze, or move the helper outside `gosplan/env/` - are the lead's to record in
`spec/CHANGELOG.md`. Until that entry exists, do not pick one and do not call `numpy.random` here: if
`tests/unit/test_prices.py` forces the question, **file an AMBIGUITY REPORT and stop.**

## Forbidden

- Any reward term that CONTRACT rule 4 does not name: no per-step shaping, no auxiliary reward, no
  curiosity term, no potential-based term, no bonus for truthfulness, no penalty for volatility. The
  five terms above are the whole reward.
- Running reward normalisation of any kind - no `RunningMeanStd`, no moving average, no
  standardisation against episode statistics. Running statistics change the effective reward over
  training and, with heavy-tailed penalties, shrink the notch in normalised units. Per-batch
  advantage normalisation lives in the PPO adapter (WO-017), not here.
- Reading `welfare_true` or `val_measured` in any reward path, or letting either reach an observation
  or an agent input (CONTRACT rule 6). They are logged quantities only.
- Smoothing the `w = 0` Heaviside, adding an epsilon to it, or clipping when `rho_cap = inf`.
- Widening or narrowing `rho_max`, or suppressing the at-bound record (CONTRACT rule 8: bounds are
  results; above 1 per cent of reports at the bound the run is flagged `BOUND_BINDING` and reported
  with the flag).
- Any direct `numpy.random` or `jax.random` call (CONTRACT rule 9); the `auditnoise` draw goes
  through `gosplan.rng.draw`.
- Importing scipy, pandas, jax or torch, at module scope or inside a function - including
  `scipy.optimize` for the price fixed point. numpy alone is enough.
- Implementing any pathology directly (CONTRACT rule 7): nothing here may detect padding, reward
  honesty, or respond to a report's history.
- Changing any signature, module constant, or module docstring; editing any test, `spec/spec.py`, or
  any module not on the "Write only" list.

## Contract rules that bite here

- **Rule 3 - STOP AND REPORT.** Choosing "the reasonable default" is a violation: do not smooth a
  discontinuity, do not add a stabilising term, do not pick an RNG purpose the enumeration does not
  contain. `perturbed_price_vectors` is an explicit open item - stop, do not resolve it.
- **Rule 12 - WORK ORDERS (whitelist).** Read only the fourteen entries above, write only the three
  files, run the completion command verbatim.
- **Rule 4 - REWARD TERMS.** The reward is exactly `-scale * c_ik` at PRODUCE and
  `scale * (B(rho) - 1[audited]*Pen + trade_surplus)` at REPORT, with `scale = reward_scale(cfg)`
  computed analytically. No shaping, no auxiliary term, no running normalisation.
- **Rule 6 - WELFARE BLINDNESS.** `welfare_true` and `val_measured` are logged and never enter an
  observation, a reward or an agent input.
- **Rule 8 - BOUNDS ARE RESULTS.** `report_ratio` is bounded at `rho_max = 10`; the at-bound record
  must be kept, not suppressed, and the bound is never silently moved.
- **Rule 9 - RNG.** The `auditnoise` draw goes through `gosplan.rng.draw` and is made
  unconditionally so the key layout is configuration-independent.
- **Rule 7 - NO HARD-CODED PATHOLOGY.** Nothing here detects padding or rewards honesty.
- **Rule 2 - FROZEN TESTS.** The three test files are read-only; disagreement is an AMBIGUITY REPORT.

## Definition of done

- [ ] `tests/unit/test_reporting.py` (T-U8) green.
- [ ] `tests/unit/test_reward.py` (T-U2, T-U3, T-B6 components) green.
- [ ] `tests/unit/test_prices.py` green.
- [ ] `uv run ruff check . && uv run ruff format --check .` clean.
- [ ] `uv run python scripts/contract_guard.py --all` clean.
- [ ] No file outside "Write only" touched.
- [ ] CI `ci-ok` green on the PR.
- [ ] The four-part report posted as the PR body: 1) tests passed/failed with output; 2) ambiguity
      reports filed, each listed, or "none"; 3) files changed - expect exactly the three files under
      "Write only"; 4) anything you were tempted to invent and did not - in particular the
      `B_cfg(1.1) = 0` case, the zero-consumption CES case, and anything `perturbed_price_vectors`
      demanded.
